### ISSUE TITLE
Allow trailing commas in shader language

### ISSUE THREAD

**Issue by akx** [godotengine/godot-proposals#8840]
### Describe the project you are working on

A shader-heavy video sequence generator

### Describe the problem or limitation you are having in your project

Long shading expressions like
```glsl
vec3 fadeInTex = clamp(mix(-1.0, 2.0, fadeIn) + texture(fade_in_bias, UV).rgb, 0.0, 1.0);
```
would be nicer formatted as
```glsl
vec3 fadeInTex = clamp(
  mix(-1.0, 2.0, fadeIn) + texture(fade_in_bias, UV).rgb, 
  0.0, 
  1.0,
);
```
but the presence of a trailing comma in the argument list (as allowed in GDScript in 4.0+, https://github.com/godotengine/godot/issues/41066) yields

> error(69): Expected expression, found: 'PARENTHESIS_CLOSE'

For context, function definition argument lists already accept a single trailing comma; both

```glsl
vec3 neonChevrons(vec2 uv,) {
```
and
```glsl
vec3 neonChevrons(
	vec2 uv,
) {
```
parse fine.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This is somewhat tautological, but:

Fixing the shading language parser to allow trailing commas in function call argument lists in the shader language parser would allow one to use trailing commas in function call argument lists.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When the user reformats their shader code to a form where a function call has a trailing comma in its argument list, the comma token would be swallowed, and no error for a missing expression would be emitted.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can not be worked around with script, because it's part of the shading language parser: https://github.com/godotengine/godot/blob/84e205b5a17bfe7ace96b78c410ec10aa52241d2/servers/rendering/shader_language.cpp#L5739-L5743

### Is there a reason why this should be core and not an add-on in the asset library?

The shading language parser is part of Godot core.

**Comment by Mickeon** [comment]
Makes sense to me. GDScript allows trailing commas in function arguments, as well (It does not in 3.x I believe).

**Comment by Calinou** [comment]
This is a good idea in principle, but keep in mind standard GLSL syntax doesn't allow this. As a result, allowing trailing commas will result in the code not being usable outside of Godot shaders without being modified.

Unlike GDScript, the Godot shader language is designed for a high level of interoperability with GLSL, so this change may have negative effects here.

**Comment by AThousandShips** [comment]
That's the wrong direction of conversion, what Calinou is talking about is taking GDShader code and using it in general GLSL shaders, which is normally possible with minor changes, as it largely doesn't have features not in GLSL

**Comment by akx** [comment]
@AThousandShips Er, well, Calinou's message does say "copy shader code between GLSL and Godot shaders **in both directions**", emphasis mine.

**Comment by AThousandShips** [comment]
Yes, but:
> As a result, allowing trailing commas will result in the code not being usable outside of Godot shaders without being modified.

The direction of relevance was Godot -> GLSL 🙂

**Comment by clayjohn** [comment]
> As mentioned above,
> 
> ```glsl
> float foo(float a,) {
> 	return a + 5.0;
> }
> ```
> 
> is valid Godot shader code (v4.2.1.stable.official [b09f793f5]), but not valid GLSL, so there's already a break there.

Honestly, this looks like a bug to me. If it works, its a happy accident and doesn't seem intentional.

**Comment by AThousandShips** [comment]
It looks like an oversight to me, since it's just for declarations, and the part of the code that handles the commas doesn't make any mention of allowing trailing comma which I'd assume it would if it was intentional

**Comment by AThousandShips** [comment]
My suggestion would be to fix the trialing comma in declarations, possibly with a deprecation clause allowing it without `DISABLE_DEPRECATED`

**Comment by akx** [comment]
I'm still wondering whether the strict parity with GLSL, one way or the other, is the pragmatic choice... There's already [this document about porting GLSL to Godot shaders](https://docs.godotengine.org/en/stable/tutorials/shaders/converting_glsl_to_godot_shaders.html), which says

> Godot uses a shading language based on GLSL **with the addition of a few quality-of-life features.**

– I think trailing commas definitely are a quality-of-life feature...

**Comment by AThousandShips** [comment]
I wouldn't agree that unusual syntax that differs from the language it is based on is a question of QOL, the same could be said for many syntax and style quirks

**Comment by AThousandShips** [comment]
Further inconsistency:
```
float arr[3] = { 1.0, 2.0, 3.0, };
```
Is also disallowed, as is:
```
MyStruct my_struct = MyStruct(foo, bar, );
```
Which uses a distinct call, as is:
```
uniform float my_uniform : hint_range(0, 1, );
```
So this applies to most cases

**Comment by akx** [comment]
To bring @AThousandShips's comment https://github.com/godotengine/godot/pull/86991#issuecomment-1887473439 into this conversation and not just that PR:

> seems glsl allows trailing comma in arrays

so there's another break to GLSL ES 3.0 compatibility here (since Godot doesn't allow those at present).

Right now there's no formal specification for the shading language that I know of, just (paraphrased) "similar to GLSL ES 3.0 with some QOL features" and the C++ implementation is the living spec.

I'd personally like to work with a slightly more pragmatic language than GLSL; for instance, automatic casting of arguments for built-in functions like `clamp()`, `vec*()`, etc. would probably be pretty useful, and so on, but that's another discussion and will require someone to have made the decision-in-principle whether breaking from GLSL is okay at all.

I understand that it's nice to be able to be able to copy Godot shader code to GLSL contexts (and even then, you need to go through the code and rename uniforms and so on), but it feels to me like strict 1:1 compatibility shouldn't be **the** priority.

**Comment by AThousandShips** [comment]
While absolute compliance doesn't have to be a priority, sure, that doesn't equate to adding complex code with possible points of failure though, so the validity of supporting this in a wider context, both from style, necessity, and cost should be the focus, unless we do argue that making a quite major departure from the style and syntax of GLSL is not a good idea (which is what I argue)
